File: tests/algorithms/x/Java/dynamic_programming/iterating_through_submasks.java
public class Main {

    static int bitwise_and(int a, int b) {
        int result = 0;
        int bit = 1;
        int x = a;
        int y = b;
        while (x > 0 || y > 0) {
            int abit = Math.floorMod(x, 2);
            int bbit = Math.floorMod(y, 2);
            if (abit == 1 && bbit == 1) {
                result = result + bit;
            }
            x = Math.floorDiv(x, 2);
            y = Math.floorDiv(y, 2);
            bit = bit * 2;
        }
        return result;
    }

    static int[] list_of_submasks(int mask) {
        if (mask <= 0) {
            throw new RuntimeException(String.valueOf("mask needs to be positive integer, your input " + _p(mask)));
        }
        int[] all_submasks = ((int[])(new int[]{}));
        int submask = mask;
        while (submask != 0) {
            all_submasks = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(all_submasks), java.util.stream.IntStream.of(submask)).toArray()));
            submask = bitwise_and(submask - 1, mask);
        }
        return all_submasks;
    }
    public static void main(String[] args) {
        System.out.println(_p(list_of_submasks(15)));
        System.out.println(_p(list_of_submasks(13)));
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Kotlin/dynamic_programming/iterating_through_submasks.kt
import java.math.BigInteger

fun panic(msg: String): Nothing { throw RuntimeException(msg) }

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun bitwise_and(a: Int, b: Int): Int {
    var result: Int = (0).toInt()
    var bit: Int = (1).toInt()
    var x: Int = (a).toInt()
    var y: Int = (b).toInt()
    while ((x > 0) || (y > 0)) {
        var abit: Int = (Math.floorMod(x, 2)).toInt()
        var bbit: Int = (Math.floorMod(y, 2)).toInt()
        if ((abit == 1) && (bbit == 1)) {
            result = result + bit
        }
        x = x / 2
        y = y / 2
        bit = bit * 2
    }
    return result
}

fun list_of_submasks(mask: Int): MutableList<Int> {
    if (mask <= 0) {
        panic("mask needs to be positive integer, your input " + mask.toString())
    }
    var all_submasks: MutableList<Int> = mutableListOf<Int>()
    var submask: Int = (mask).toInt()
    while (submask != 0) {
        all_submasks = run { val _tmp = all_submasks.toMutableList(); _tmp.add(submask); _tmp }
        submask = bitwise_and(submask - 1, mask)
    }
    return all_submasks
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(list_of_submasks(15).toString())
        println(list_of_submasks(13).toString())
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
